### ISSUE TITLE
Save game sharing and replays

### ISSUE THREAD

**Issue by toddmedema** [toddmedema/electrify#109]
0. DONE: Utilize random seed https://github.com/toddmedema/electrify/issues/78
1. DONE: Store game state to (local storage? firebase?), and re-load to exact same state 
2. Import/export JSON blob of game state so you can share saves
3. Store and view replay: Initial random seed + list of decisions made + current state (and then replay you just re-watch the player actions at their realtime)

**Comment by toddmedema** [comment]
## Implementation plan

Notes from reading the current code, plus a phased plan for the three items above.

### Context: what's already there vs. what's missing

#78 is closed and left real groundwork: `GameType.seed` exists, `splitmix32` lives in `src/helpers/Math.tsx:25`, and `src/reducers/Game.tsx:165` seeds it in `initGame`. There's even a `// TODO load game state from localstorage if loading` sitting at exactly the right spot — `src/components/views/LoadingContainer.tsx:43`.

But the seed alone does **not** currently determine the run, and that's the crux of this issue. Three module-level mutable caches sit outside Redux and outside any save blob:

| State | Where | Problem for load |
|---|---|---|
| `prng` closure | `src/helpers/Math.tsx:37` | Sequential — its value depends on how many draws happened, so it can't be reconstructed |
| `weather` array | `src/data/Weather.tsx:14` | CSV rows 1980–2019 **plus** PRNG-generated rows appended by `forecastNextDay` |
| `fuelPrices` | `src/data/FuelPrices.tsx:37` | Future years generated lazily by PRNG |

Two latent bugs make this worse, and both surface only on load:

- **`src/data/Weather.tsx:82`** — `getWeather` fills one month per cache miss. Load a year-2050 save into a fresh array (CSV ends 2019) and every lookup returns `DUMMY_WEATHER` (0 °C, 10 kph, no clouds) for ~370 game-months until it catches up.
- **`src/data/FuelPrices.tsx:88`** — `getFuelPricesPerMBTU` walks back to the nearest cached year. On a cold cache in 2050 it jumps straight from the last CSV year to 2050, skipping the 30-year multiplicative chain, so prices land nowhere near where the live game had them.

So a naive "JSON.stringify the game slice" save will load into visibly different weather and prices. Phase 0 fixes that at the root.

---

### Phase 0 — Make the seed actually determine the run

Prerequisite for all three numbered items, and what makes item 3 (replay) possible at all.

**Replace the sequential PRNG with an index-addressed one.** In `src/helpers/Math.tsx`:

```ts
// Pure: same (seed, stream, index) always yields the same value, in any order,
// from any process, with no accumulated state to save.
export function randomAt(seed: number, stream: number, index: number): number
export function getRandomRangeAt(seed, stream, index, min, max): number
```

Implement by hashing `seed`, `stream`, and `index` together (splitmix32-style mixing of the three, or seed once per `(stream, index)` pair and take one draw). Keep `seedRandom`/`getRandomRange` as thin wrappers during migration, then delete them — there are only **two** production call sites (`src/data/Weather.tsx:148-150`, `src/data/FuelPrices.tsx:109`), so this is a small refactor with an outsized payoff.

**Rework the two caches to be pure functions of the seed.** `DAYS_PER_MONTH = 1`, so a weather "day" maps 1:1 onto a game month — index it as `dayIndex = (year - 1980) * 12 + (month - 1)`:

- `forecastNextDay()` → `forecastDay(seed, dayIndex)`, drawing its three modifiers from `getRandomRangeAt(seed, WEATHER_STREAM, dayIndex * 3 + n)`.
- `getWeather` gains a **forward fill**: generate every missing `dayIndex` from end-of-CSV up to the requested date, in order. Fixes the `DUMMY_WEATHER` bug and makes a cold cache identical to a warm one.
- `getFuelPricesPerMBTU` likewise **chains forward** year by year from the last CSV year instead of jumping, drawing each multiplier from `(seed, FUEL_STREAM, (year * 12 + month) * 4 + fuelIdx)`.

**Store integer seeds.** `src/reducers/Game.tsx:99` and `:165` use `Date.now() * Math.random()`, a float that `splitmix32` silently truncates via `a |= 0`. The persisted seed therefore doesn't round-trip honestly — a real problem once seeds are shared in exported saves. Use `Math.floor(Math.random() * 2 ** 32)`.

**The payoff:** after this phase, a save needs to carry *only* the game slice. Both caches rebuild themselves on demand, identically, from the seed alone. Zero extra bytes, and no risk of the caches drifting out of sync with the save format later.

**Phase 0 test (the one that matters):** using the existing headless `src/testing/Simulator.tsx`, run 24 months → serialize → reset all module caches → deserialize → run 24 more, and assert `monthlyHistory` matches an uninterrupted 48-month run **exactly**. If that passes, save/load is correct by construction rather than by inspection.

---

### Phase 1 — Save/load to localStorage (item 1)

**Storage choice: localStorage, not Firebase.** It works logged-out (Firebase auth is optional here), needs no Firestore rules changes, and is synchronous. Firestore's 1 MB document limit is also uncomfortably close to a long game's save. Cloud sync keyed on `user.uid` is a sensible follow-up, not a v1 requirement.

**Save shape** — new `src/SaveGame.tsx`:

```ts
export interface SaveGameType {
  version: number;      // bump on any breaking schema change; mismatched saves are ignored, not migrated
  savedAt: string;
  appVersion: string;   // package.json, for bug reports
  game: GameType;       // the whole slice
}
```

Save the **whole** `game` slice, including `timeline`. It's tempting to drop `timeline` as derived data, but current cash and customers live inside it (`getTimeFromTimeline(...).cash`), not in scalar fields — dropping it loses the player's money.

**Size estimate:** ~96 timeline entries (~29 KB) + one `monthlyHistory` entry per game-month (~250 B each) + facilities (~400 B each, mostly the `description` strings). A 30-year scenario lands around 110 KB; a 100-year run around 350 KB. Comfortable against a ~5 MB budget. `checkStorageFreeBytes` (`src/LocalStorage.tsx:96`) already exists to verify this at runtime rather than trusting my arithmetic.

**New `loadGame` reducer** in `src/reducers/Game.tsx`. Beyond assigning the slice, it must reset the module-level tick-loop locals (`src/reducers/Game.tsx:85-97`), which `initGame` already partly does:

- `previousMonth = state.date.month` — otherwise the load immediately triggers a spurious month rollover, double-recording history.
- `previouslyInBlackout` — derive from the restored timeline so the toast doesn't misfire.
- `tickLoopRunning = false`, `previousTickMs = performance.now()`, `speedBeforeDialog = "PAUSED"`.
- Force `state.speed = "PAUSED"` so the sim doesn't run before the player is looking.

**Autosave** via `store.subscribe` in `src/Store.tsx` — keeps the write out of the reducer. Trigger on game-month change while `game.inGame`, plus on `visibilitychange`/`beforeunload`. One write per game-month is roughly one every few seconds at FAST speed, so throttle it. **Skip tutorials** (`scenario.tutorialSteps`) — they're short and restoring mid-Joyride `tutorialStep` is not worth the complexity.

**Load path** — `src/components/views/LoadingContainer.tsx:43`, where the TODO already is. After `initWeather` and `initFuelPrices` resolve, branch: dispatch `loadGame(save)` instead of `initGame({...})`, then `loaded()` as usual.

**UI:**
- "Continue" button in `src/components/views/MainMenu.tsx`, above Play, rendered only when a valid save exists, subtitled with scenario name and in-game year.
- Confirm dialog before starting a new game when a save exists (it would be overwritten).
- Clear the save when a scenario ends — bankrupt/fired/win, the `dialogOpen` triggers around `src/reducers/Game.tsx:378`. Quitting to the menu should *keep* it.
- Failed write (quota) → snackbar. `setStorageKeyValue` swallows errors by default; pass `ignoreErrors: false` here and catch.

**Tests:** round-trip preserves cash/facilities/history; a version mismatch is ignored without crashing; corrupt JSON is ignored (`getStorageJson` already returns the fallback); loading doesn't double-record a month.

---

### Phase 2 — Import/export JSON (item 2)

Add to `src/components/views/Settings.tsx` (and optionally the in-game menu):

- **Export** — serialize `SaveGameType`, `Blob` + object URL + `<a download>`, filename `electrify-<scenario>-<year>.json`.
- **Import** — `<input type="file">`, parse, validate `version` and a handful of required fields, reject with a clear message otherwise, then route through the same `loadGame` path.

Because Phase 0 made the seed authoritative, a shared save reproduces the donor's exact weather and price trajectory on the recipient's machine. Worth validating strictly here — an imported blob is untrusted input, so check shape before it reaches the reducer rather than letting a malformed facility crash the sim mid-tick.

---

### Phase 3 — Replay (item 3)

Every state mutation already flows through a Redux action, which makes this mostly mechanical.

**Record** — a small middleware appending `{ minute: state.game.date.minute, action }` for a whitelist: `buildFacility`, `sellFacility`, `togglePauseFacility`, `reprioritizeFacility`, and the `delta` payloads that matter (`dollarsPerkWh`, `monthlyMarketingSpend`, `difficulty`). `setSpeed` is *not* simulation-affecting — `tickState` always advances a fixed `TICK_MINUTES = 15` regardless of wall-clock speed — so record it only if you want playback to mimic the player's pacing.

**Replay file:** `{ version, seed, scenarioId, difficulty, actions: [...] }`. Tiny — kilobytes, not hundreds of kilobytes.

**Playback:** `initGame` with the recorded seed, then loop `tickState`, dispatching each recorded action when `date.minute` reaches its stamp. Deterministic only because of Phase 0.

This also gives a strong regression harness for free: replay a recorded game against a new build and diff the monthly history to catch unintended balance changes.

---

### Suggested sequencing

Phase 0 is the only phase with a hard ordering constraint, and it's the one carrying the real risk — it changes simulation output for a given seed, so any stored baseline numbers in `src/testing/Simulation.test.tsx` will need regenerating. Phases 1–3 map 1:1 onto the numbered items above and can ship independently, each behind its own PR.

Rough sizes: Phase 0 is ~200 lines across three files plus the determinism test; Phase 1 is the bulk of the UI work; Phase 2 is small; Phase 3 is self-contained.
